### PR TITLE
[WIP] Harden systemd unit file

### PR DESCRIPTION
**Pull request by paulfantom**
##### Summary

Update systemd file with a more secure version

Closes #1671

##### Component Name

Installer, service file

##### Additional Information

Rebased and updated version of #1671. 

```paste below

```

### CODE REVIEW

File: system/netdata.service.in [hunk 1]
```diff
@@ -1,31 +1,37 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
 [Unit]
 Description=Real time performance monitoring
-
+#Documentation=man:netdata
+#Documentation=file:///usr/share/doc/netdata/html/index.html
+Documentation=https://github.com/netdata/netdata/wiki
```
Comment (ktsaou) [inline system/netdata.service.in]: This should point to `https://docs.netdata.cloud`.

File: system/netdata.service.in [hunk 2]
```diff
@@ -1,31 +1,37 @@
...
 # append here other services you want netdata to wait for them to start
 After=network.target httpd.service squid.service nfs-server.service mysqld.service mysql.service named.service postfix.service chronyd.service
 
 [Service]
 Type=simple
 User=netdata
 Group=netdata
+Environment="netdata_LOG_LOCATION=/var/log/netdata/log"
```
Comment (ktsaou) [inline system/netdata.service.in]: Why is this needed?
If we need it, it should be `logdir_POST@`

File: system/netdata.service.in [hunk 3]
```diff
@@ -37,12 +43,28 @@ OOMScoreAdjust=1000
...
 CPUSchedulingPolicy=idle
 
-# This sets the scheduling priority (for policies: rr and fifo).
-# Priority gets values 1 (lowest) to 99 (highest).
-#CPUSchedulingPriority=1
+NoNewPrivileges=false
+PermissionsStartOnly=true
+# CAP_SETGID is required for setgroups()
+# CAP_NET_RAW is needed by fping
+CapabilityBoundingSet=CAP_DAC_READ_SEARCH CAP_SYS_PTRACE CAP_SETGID CAP_NET_RAW
```
Comment (ktsaou) [inline system/netdata.service.in]: Why is this needed?
netdata daemon does not need any capabilities.

File: system/netdata.service.in [hunk 4]
```diff
@@ -37,12 +43,28 @@ OOMScoreAdjust=1000
...
+PrivateTmp=true
+ProtectHome=read-only
+ProtectSystem=full
+
+ReadOnlyDirectories=/
```
Comment (ktsaou) [inline system/netdata.service.in]: If netdata sees all mount points as read-only it will not report disk space usage (size, inodes).
This setting actually kills `diskspace.plugin`.

I am not sure what is right to do in this case. Leave the directories read-write or remove the detection of read-only mount points and report disk space usage even on read-only ones.

**Comment by ktsaou** [comment]
I made a few comments.

In general, netdata is already running unprivileged (not requiring any capabilities, not running as root). But at the same time it needs to have a good view of the system to be able to collect and report its performance metrics.

So, I am not sure the proposed systemd service file is actually "hardened". It seems like it gives unnecessary capabilities to netdata while restricting its view on the system.

**Comment by FedericoCeratto** [comment]
@ktsaou please keep in mind that the unit file sets capability bounding (not capabilities) and filesystem sandboxing for all processes run by Netdata. Some capabilities are/were required by apps.plugin, and are explicitly set on the binary e.g. :
https://salsa.debian.org/debian/netdata/blob/master/debian/netdata.postinst.in#L52

It would be ideal to provide more fine-grained sandboxing by splitting the main netdata process, the built-in webserver and the collector(s) in different services.
This should perhaps be discussed in a dedicated (long-term) issue around Netdata's security model, what do you think?

**Comment by ktsaou** [comment]
> It would be ideal to provide more fine-grained sandboxing by splitting the main netdata process, the built-in webserver and the collector(s) in different services.
> This should perhaps be discussed in a dedicated (long-term) issue around Netdata's security model, what do you think?

The same has been proposed by @alonbl since the very beginning. But I find it very counter intuitive from a usability perspective. There are 2 plugins running escalated (apps, freeipmi) and a few more requiring `sudo`. Out of the dozens that run fine unprivileged.

Even if we split netdata into multiple services, there will be cases this will not entirely fix the problem, like `cgroup-network` (that is setuid to root, to find the network interfaces linked to containers).

If we re-evaluate the whole security model, I would probably go with a daemon running as root, spawning unprivileged processes for all its functions. This could allow multi-container monitoring like it is described at https://github.com/netdata/netdata/issues/474#issuecomment-223142222

Probably I am missing something. Can you help me understand how netdata will be more secure by running privileged plugins as independent systemd services?

**Comment by alonbl** [comment]
A quick answer... using root should be deprecated in these days, I understand this is the easy method and the method you have gotten used to in the past... but... you must drop this approach in favor of running completely unprivileged components each with least privilege capabilities and communicate with each other via RPC. This approach supports restriction of the damage domain that your application can create.

The question left to be discussed - what is the privilege component that execute your application... There are options such as init.d system (systemd, openrc, etc...) or orchestrators such as k8s.

Let me repeat... assuming root in any application is a fault in design for modern operating systems.

**Comment by paulfantom** [comment]
I completely agree with @alonbl.

**Comment by FedericoCeratto** [comment]
As discussed with @paulfantom on IRC, sandboxing the entire netdata process with systemd is simple but very coarse-grained. An ideal approach would be to spawn every plugin with its own custom sandboxing.

**Comment by cakrit** [comment]
cc @paulkatsoulakis

**Comment by paulfantom** [comment]
Closing as I won't be working on it and this is on my fork.
